Add the "Add another item" action for all collections

thomasbeaujean opened the pull request:
Use case:
I have an entity that has a property that has a collection. This collection has also a collection.
Ex: Entity->Collection1->Collection2

So I expected to have to "Add another Item" for the collection of level 1 and 2. But only the first level displayed the "add" button.

The "add button" was not handled within the symfony widgets, that is why the behaviour was not generic.

I moved the JS and the html from Resources/views/default/form.html.twig to Resources/views/form/bootstrap_3_horizontal_layout.html.twig

Pierstoval commented:
I wish there could be a proper phpunit test to check whether the collection button appears correctly for nested collections :confused:

thomasbeaujean commented:
I thought that my modification was enough but it appears I still got a JS issue. The JS is still not generic enough for several nested levels.
Please do not consider this PR while I do not have fully check this issue.
@Pierstoval Ok, I will try to add a test for the main usage

ogizanagi commented:
Great ! Indirectly move another step forward to https://github.com/javiereguiluz/EasyAdminBundle/issues/499#issuecomment-151653895 :)

Pierstoval commented:
Plus, shouldn't this code be written also in `bootstrap_3_layout.html.twig`? You might consider putting this feature inside an include, or even in a twig macro.

ogizanagi commented:
And part of the `collection_row/widget` instead of `form_row` instead ?

javiereguiluz commented on Resources/views/form/bootstrap_3_horizontal_layout.html.twig:
```diff
@@ -36,6 +36,30 @@ col-sm-2
                 <span class="help-block"><i class="fa fa-info-circle"></i> {{ form.vars.attr.field_help|trans|raw }}</span>
             {% endif %}
         </div>
+        {% if form.vars.allow_add|default(false) %}
```
According to @ogizanagi we can simplify `form.vars.allow_add` as `allow_add`

javiereguiluz commented on Resources/views/form/bootstrap_3_horizontal_layout.html.twig:
```diff
@@ -36,6 +36,30 @@ col-sm-2
             {% endif %}
         </div>
+        {% if form.vars.allow_add|default(false) %}
+            {% set add_item_javascript %}
```
I neverliked the name of this variable (`add_item_javascript`) because it looks like you are adding an item called "javascript". What about `js_add_item`, `add_item_js`, `js_code_add_item` or something like that. Any other proposal?

thomasbeaujean commented:
Thank you for these good advices.

As suggested:
* I renamed the add_item_javascript to js_add_item 
* I add block collection_row
* I put the code in bootstrap_3_layout.html.twig

And I finally could manage to fix my js issue.

# The identifiers
So if you have two nested forms, the strings to replace are like the following:

*  __ name__label__ 
* some_form_entityOne_0_entityTwo___name__
* commitment_type_form[entityOne][0][entityTwo][0][name]

## The previous behaviour

You add an element, the __ name __ are replaced by the the item index.
So __ name __ is replaced by 0.

You want to add a sub item, the  some_form_entityOne_0_entityTwo___name__ has already been replaced by some_form_entityOne_0_entityTwo_0, and in the js.

I want to add a second sub item, and all the 0 are replaced by 0, I get a some_form_entityOne_1_entityTwo_1 instead of some_form_entityOne_0_entityTwo_1.

In the second level case, the __ name __ is the parent index (0, 1 etc.), so before, the code tried to replace the parent index by the child index (replace 0 by 1 for example).
The first level of nested form had already replace all the __ name __ by its index.

# The solution
To avoid this behaviour, I modified the "replace" in order to be more specific in the search using the name of the property.

## The modification in the js

+ .replace(/\>__ name__label__\</g, '>' + numItems + '<')
+ .replace(/_{{ form.vars.name}}___name__/g, numItems)
+ .replace(/{{ form.vars.name}}\]\[__name__\]/g, '{{ form.vars.name}}][' + numItems + ']')

I hope my explanation is clear enough.

ogizanagi commented on Resources/views/form/bootstrap_3_layout.html.twig:
```diff
@@ -273,3 +273,33 @@
+            $(function() {
+                if(event.preventDefault) event.preventDefault(); else event.returnValue = false;
+
+                var collection = $('#{{ form.vars.id }}');
```
```diff
- var collection = $('#{{ form.vars.id }}');
+ var collection = $('#{{ id }}');
```

and same for other occurrences below. :)

javiereguiluz commented:
@thomasbeaujean thanks for improving this code. I like it a lot because it simplify templates and it puts things where they should be. Thanks!
